src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp

```diff
@@ -0,0 +1,901 @@
+            _p2match.erase(it);
+            return;
+        }
+    }
+    
+}
+
+std::string PDPServer::GetPersistenceFileName()
+{
+    assert(getRTPSParticipant());
+
+   std::ostringstream filename(std::ios_base::ate);
+   std::string prefix;
+
+   // . is not suitable separator for filenames
+   filename << "server-" << getRTPSParticipant()->getGuid().guidPrefix;
+   prefix = filename.str();
+   std::replace(prefix.begin(), prefix.end(), '.', '-');
+   filename.str(std::move(prefix));
+   filename << ".db";
+
+   return filename.str();
+
+}
+
+bool PDPServer::all_servers_acknowledge_PDP()
+{
+    // check if already initialized
+    assert(mp_PDPWriterHistory && mp_PDPWriter);
+
+    // First check if all servers have been discovered
+    bool discovered = true;
+
+    for (auto & s : mp_builtin->m_DiscoveryServers)
+    {
+        discovered &= (s.proxy != nullptr);
+    }
+
+    if (!discovered)
+    {
+        // The first change in the PDP WriterHistory is this server ParticipantProxyData
+        // see BuiltinProtocols::initBuiltinProtocols call to PDPXXX::announceParticipantState(true)
+        CacheChange_t * pPD;
+        if (mp_PDPWriterHistory->get_min_change(&pPD))
+        {
+            // This answer includes also clients but is accurate enough
+            return mp_PDPWriter->is_acked_by_all(pPD);
+        }
+        else
+        {
+            logError(RTPS_PDP, "ParticipantProxy data should have been added to client PDP history cache by a previous call to announceParticipantState()");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+
+bool PDPServer::is_all_servers_PDPdata_updated()
+{
+    StatefulReader * pR = dynamic_cast<StatefulReader*>(mp_PDPReader);
+    assert(pR);
+
+    // This answer includes also clients but is accurate enough
+    return pR->isInCleanState();
+}
+
+
+bool PDPServer::match_servers_EDP_endpoints()
+{
+    std::lock_guard<std::recursive_mutex> lock(*getMutex());
+    bool all = true; // have all servers been discovered?
+
+    for (auto & svr : mp_builtin->m_DiscoveryServers)
+    {
+        all &= (svr.proxy != nullptr);
+
+        if (svr.proxy && !mp_EDP->areRemoteEndpointsMatched(svr.proxy))
+        {
+            queueParticipantForEDPMatch(svr.proxy);
+        }
+    }
+
+    return all;
+}
+
+void PDPServer::announceParticipantState(bool new_change, bool dispose /* = false */, WriteParams& )
+{
+    StatefulWriter * pW = dynamic_cast<StatefulWriter*>(mp_PDPWriter);
+    assert(pW);
+
+    // Servers only send direct DATA(p) to servers in order to allow discovery
+    if (new_change)
+    {
+        // only builtinprotocols uses new_change = true, delegate in base class
+        // in order to get the ParticipantProxyData into the WriterHistory and broadcast the first DATA(p)
+
+        WriteParams wp;
+        SampleIdentity local;
+        local.writer_guid(mp_PDPWriter->getGuid());
+        local.sequence_number(mp_PDPWriterHistory->next_sequence_number());
```

Comment:
`next_sequence_number` is non-thread-safe.

Reply:
Agreed!
I introduce a lock to enforce atomic access.

Reply:
Possible deadlock. Make sure we always lock in the same order: first the PDP mutex and then the PDP writer mutex.